add rstudio-desktop recipe

izahn opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/rstudio-desktop```) and found it was in an excellent condition.

izahn commented:
@h-vetinari are you still interested in packaging rstudio? Would love to join forces. In contrast to https://github.com/conda-forge/staged-recipes/pull/13760 this PR takes an incremental approach. Rather than try to do everything at once I started with getting a working linux x86 desktop recipe. Once we get that to a known good state we can build on it, e.g. by building for other platforms and architectures. 

Right now this PR does produce a working rstudio build for x86 linux, so we are off to a good start. I'm working locally now to clean up a couple of things, and then I think this will be ready to merge. I'll be extremely happy for any help! I'm especially looking for help with future OSX / Windows builds since I don't have either to build or test on locally.

h-vetinari commented:
Hey @izahn 

Great that you're tackling this, and yes, I'd still be interested - I'd imagine @thomann might be as well?

The (most immediate) problem in #13424 was that we were stuck on the discussion of how to deal with PAM. It would be possible to add a CDT (cf. https://github.com/conda-forge/cdt-builds/pull/55), which is what anaconda did (and I tend to agree, packaging such low-level security-critical software is a tall order that I'm really not sure we should take on), but there were voices that this should be attempted, which I did in https://github.com/conda-forge/staged-recipes/pull/16911. Then the discussion petered out unfortunately.

izahn commented:
Ready for review @conda-forge/help-r @conda-forge/help-c-cpp

h-vetinari commented:
Given that it's the same sources, I still think that -server and -desktop should be in the same recipe. But this can also be tackled later on the feedstock if necessary.

I think it would also be good to test explicitly that the conda-forge-packaged RStudio really properly picks up the R installation from the environment.

izahn commented:
Definitely agree about rstudio server, I'll take a look at it soon.

izahn commented:
I haven't figured out a way to do automated testing with rstudio, but I have installed this package locally and manually tested it. It does pick up the correct R installation.

izahn commented:
I tried building the server, and ran smack into 
```fatal error: security/pam_appl.h: No such file or directory
   19 | #include <security/pam_appl.h>
```
so this needs the pam cdt you mentioned @h-vetinari .

Since the desktop package doesn't need that I suggest getting the desktop build up and running and then working on the server over time. Personally I would prioritize desktop packages for OSX and Windows first, but of course I'll be delighted if someone wants to pick up on https://github.com/conda-forge/cdt-builds/pull/55 and push the server build forward.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/rstudio```) and found it was in an excellent condition.

izahn commented on recipes/rstudio/meta.yaml:
```diff
@@ -0,0 +1,155 @@
+        - r-rstudioapi
+        - libxkbcommon  # [linux]
+        - openjdk
+        - boost
```
Oh, I don't think so, thanks! Removed in 47f1c81ea4052473c13e88ed4c30308bf3343bb8

isuruf commented on recipes/rstudio/build-desktop.sh:
```diff
@@ -0,0 +1,78 @@
+cmake -S . -B build \
+      -DRSTUDIO_TARGET=Desktop \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_INSTALL_PREFIX="${PREFIX}/lib/rstudio" \
```
Patch https://github.com/rstudio/rstudio/blob/edd144e6c64f765f62716e57a31808597e92ab7e/src/cpp/session/include/session/SessionConstants.hpp#L143 to be just `bin` instead of `bin/pandoc`?

izahn replied:
Yeah, that will be nicer. OK, will do that instead, thanks!

izahn replied:
I have a local build that installs to `RPREFIX` instead of `PREFIX/lib/rstudio`, but now I'm not so sure this is a good idea. There is a lot of stuff in the top-level directory that doesn't belong there, e.g., `COPYING`, `INSTALL` `www`. It looks to me like the build is designed to install rstudio to a stand-alone directory as I did before. Thoughts?

isuruf replied:
defaults package also uses `PREFIX`, so it's better to use the same layout.

izahn replied:
OK I'll clean it up as best I can.

izahn replied:
Well I spent some more time on it and managed to break it quite nicely :-(

At this point I'm against trying to force it into a standard directory layout. The old defaults channel package managed this, but only with extensive patches (e.g., https://github.com/AnacondaRecipes/aggregateR/blob/be32d88d1ccc5735a34522f21f0e56dbd0849668/rstudio-feedstock/recipe/0028-Conda-Put-resources-in-share-rstudio.patch ; all together there are 45 patches needed for that recipe, and the maintenance burden of all those patches is  I think the main reason why we haven't had a working rstudio conda package for years now.

I'm happy to maintain a straight-up packing of the upstream releases, but maintaining an extensive set of local patches is more than I'm willing to do. Happy to get others thoughts on this too @conda-forge/help-r

isuruf replied:
> Well I spent some more time on it and managed to break it quite nicely :-(

What broke?

> please let me know if you don't think that is acceptable.

Without knowing how things broke I can't say if this is acceptable or not.

izahn replied:
The build completes without error, but the resulting binary does not work because it looks for resources in the wrong place. My assessment is that the only way to get a freedesktop-like layout is to extensively patch the upstream sources, which I am unwilling to do.

izahn replied:
You can see the problem in https://github.com/conda-forge/staged-recipes/pull/18401

izahn replied:
When I try to run the binary produced by https://github.com/conda-forge/staged-recipes/pull/18401 I get this:
```
TTY detected. Printing informational message about logging configuration. Logging configuration loaded from '/home/izahn/.config/kdedefaults:/etc/xdg/rstudio/logging.conf'. Logging to '/home/izahn/.local/share/rstudio/log/rdesktop.log'.
2022-03-19T14:53:08.987141Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/error.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::SessionLauncher::showLaunchErrorPage() src/cpp/desktop/DesktopSessionLauncher.cpp:379
"Uncaught TypeError: Cannot read property 'notifyRCrashed' of undefined"
2022-03-19T14:53:09.051025Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:09.069802Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:09.304886Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:09.724277Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:10.344989Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:11.163897Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:12.183333Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:13.403856Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
2022-03-19T14:53:14.827088Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /home/izahn/mambaforge/envs/rstudio/bin/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::basic_istream<char> >&) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
```
I uploaded it to my izahn channel to make it easier to install and test locally if you want.

izahn replied:
We spent some time on this in https://github.com/conda-forge/staged-recipes/pull/18401 without success. I also have concerns about the patches there causing other problems. It's this is not broken I'm not going to fix it. #wontfix

h-vetinari commented:
> @h-vetinari can I add you as a maintainer?

Sure

izahn commented:
@conda-forge/core Per the instructions at https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=macosx_sdk_version#requiring-newer-macos-sdks I'm reaching out because I think I need to set `MACOSX_SDK_VERSION=10.12` in order to build this package for osx. The problem is that without this it cannot find `faccessat`, as noted at https://github.com/util-linux/util-linux/issues/716 this is only available in more recent OSX SDKs.

izahn commented:
We have an OSX build now, but I don't have a Mac to test it on. If you have a Mac and can test the artifacts can be found in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=477502&view=artifacts&pathAsName=false&type=publishedArtifacts . I also uploaded to my `izahn` channel for convenience.

izahn commented:
Awesome, thanks! If it opens and starts an R console without error I'll be pretty happy. If running
```
plot(1:10)
```
in the console displays a graph I'll be even happier!

timsnyder commented:
![image](https://user-images.githubusercontent.com/83846/159280460-c8a78e9e-8d6b-42eb-9bc2-48be97426a36.png)

timsnyder commented:
grepping the other feedstocks for xvfb and just grabbing one that flew by, I found https://github.com/conda-forge/fsleyes-widgets-feedstock/tree/main/recipe.  It has a very trivial example of how you can run something in an `xvfb-run` wrapper that will want to pop open a graphical display.

izahn commented:
Very cool, I'll give it a try. Thanks!

jdblischak commented:
@izahn Looks to be working well for me!

```sh
mamba create -n rstest izahn::rstudio-desktop
conda activate rstest

conda list rstudio-desktop
# packages in environment at /usr/local/Caskroom/miniforge/base/envs/rstest:
#
# Name                    Version                   Build  Channel
# rstudio-desktop           2022.02.0         r41h011d700_0    izahn

rstudio
# TTY detected. Printing informational message about logging configuration. Logging configuration loaded from '/etc/rstudio/logging.conf'. Logging to '~/.local/share/rstudio/log/rdesktop.log'.
# 2022-03-21T14:35:31.552142Z [rdesktop] WARNING r-ldpath script not found at /usr/local/Caskroom/miniforge/base/envs/rstest/lib/rstudio/RStudio.app/Contents/MacOS/../session/r-ldpath; LOGGED FROM: std::string rstudio::core::r_util::(anonymous namespace)::extraLibraryPaths(const rstudio::core::FilePath &, const std::string &) src/cpp/core/r_util/REnvironmentPosix.cpp:84
# 2022-03-21T14:35:39.114099Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /usr/local/Caskroom/miniforge/base/envs/rstest/lib/rstudio/RStudio.app/Contents/MacOS/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::istream> &) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
```

![Screen Shot 2022-03-21 at 10 36 14 AM](https://user-images.githubusercontent.com/1608317/159285094-b3d88cb1-85f6-4f40-9196-f31c79e696c4.png)

izahn commented:
The xvfb suggestion has been really useful, and I was able to add use it to add a test on Linux.

I'm not sure this will work for OSX though -- AFAIK the rstudio GUI doesn't use X-windows on Mac. I'm not sure that we would expect rstudio to run under xvfb on that platform. Or do we?

timsnyder commented:
Yeah, I wasn't thinking hard enough earlier this morning.  Since rstudio is built to use the native display code on mac, xquartz won't help there.

timsnyder commented:
It looks like the rstudio folks use Electron to test on mac in CI but that they are having trouble with their own setup https://github.com/rstudio/rstudio/commit/88cf89c0a0587036408d0a6f040c637cebe78834

izahn commented:
> It looks like the rstudio folks use Electron to test on mac in CI but that they are having trouble with their own setup [rstudio/rstudio@88cf89c](https://github.com/rstudio/rstudio/commit/88cf89c0a0587036408d0a6f040c637cebe78834)

Yeah, that looks like a whole can of worms. I did add a real test using xvfb in linux in https://github.com/conda-forge/staged-recipes/pull/18320/commits/2bfe1f1a2675697971286abf3747b9cfb13a5b08 and that is already a big improvement.

isuruf commented:
Commenting here

> I feel like we're making this hard on ourselves, and I don't really see the benefit. https://github.com/conda-forge/staged-recipes/pull/18320 works (for sure on linux, maybe on osx too, I just don't have osx available to test). Why don't we just use that?

We have a standard layout for unix for packages and I'd like to use that unless there's no other way.

izahn commented:
> Commenting here
> 
> > I feel like we're making this hard on ourselves, and I don't really see the benefit. #18320 works (for sure on linux, maybe on osx too, I just don't have osx available to test). Why don't we just use that?
> 
> We have a standard layout for unix for packages and I'd like to use that unless there's no other way.

It's not impossible, but there are trade-offs. I believe it will take considerable time to figure out how to patch the sources to produce the desired outcome. These patches may break some functionality, perhaps in subtle ways, and they would have to be maintained by the feedstock maintainers.

While I agree that sticking to a free-desktop-like layout would be nice, there are plenty of examples where the conda-forge community has decided the effort isn't worth it. The `r-base` package is a good example of this, it also installs everything under `lib/R` and installs a wrapper in `bin/R` that calls the actual executable. We could re-arrange `r-base` to conform to  a standard layout, but it's probably not worth it. I believe the same is true here.

If someone wants to undertake the effort to patch and test to make sure the patches don't break anything I won't object. I tried and it was hard, and I don't think its worth it. I'm not going to be the one to do it.

timsnyder commented:
@izahn I notice that you tried setting `RSTUDIO_PACKAGE_BUILD` in https://github.com/izahn/staged-recipes/commit/77c3c688e8d5eca6213b2a78c3952afcd7b830d4 and then reverting it.  What does `RSTUDIO_PACKAGE_BUILD` do differently in the build?  I haven't looked closely yet.  I did notice that you had a lower-case `-d` on it.  How does CMAKE handle that?  

I also wonder if `CMAKE_BUILD_TYPE=Release` and `RSTUDIO_PACKAGE_BUILD` are possibly mutually exclusive only based on https://github.com/rstudio/rstudio/blob/e400bbf755174d8fb74691d22f03160d6a4b0c89/INSTALL#L38-L39

izahn commented:
The documentation of these variables is sparse, but one effect of RSTUDIO_PACKAGE_BUILD is to force building an internal version of qt and using it instead of the conda-forge one. A lot of other things go wrong when that variable is set too.

There was a lowercase d bug that I fixed in the next commit. I reverted both because for the reason noted above.

izahn commented:
If I had to guess I would say that RSTUDIO_PACKAGE_BUILD is used to build the packages at https://www.rstudio.com/products/rstudio/download/#download

timsnyder commented:
Interesting.   Looking at https://github.com/rstudio/rstudio/blob/edd144e6c64f765f62716e57a31808597e92ab7e/src/cpp/desktop/DesktopOptions.cpp#L463-L471 it seems like we should be using `RSTUDIO_PACKAGE_BUILD` (because it says it is for 'release' builds and not setting it is for 'developer' builds).

But I do see what you say about it forcing QT to be bundled in https://github.com/rstudio/rstudio/blob/fea9c7dcc9271e3eb4346e168d9ccde532944c0b/src/cpp/desktop/CMakeLists.txt#L324-L334

timsnyder commented:
It also seems to make rstudio statically link Boost into itself on linux, yuk 🤮 
https://github.com/rstudio/rstudio/blob/149b4e95f4683fef6c6e9ab3394f53979be10780/src/cpp/CMakeLists.txt#L227-L232

timsnyder commented:
I'm satisfied that you can't use `RSTUDIO_PACKAGE_BUILD` as it exists.

izahn commented:
I'm pretty well satisfied with the state of the Linux and OSX builds. The Linux build now has an actual CI test, and I have also tested it locally and everything seems to work well. @timsnyder and @jdblischak if you have time to test the OSX builds once more that would be amazing. In particular I'm hoping to have eliminated the ` WARNING r-ldpath script not found` and `ERROR system error 2 (No such file or directory) [path: <SNIP> resources/html/loading.html]` messages.

jdblischak commented:
Looks like the `r-ldpath` error is gone.

```sh
mamba create -n rstest2 izahn::rstudio-desktop
conda activate rstest2

conda list rstudio-desktop
# Name                    Version                   Build  Channel
# rstudio-desktop           2022.02.1         r41h011d700_0    izahn

rstudio
# TTY detected. Printing informational message about logging configuration. Logging configuration loaded from '/etc/rstudio/logging.conf'. Logging to '~/.local/share/rstudio/log/rdesktop.log'.
# 2022-03-26T00:10:51.798047Z [rdesktop] ERROR system error 2 (No such file or directory) [path: /usr/local/Caskroom/miniforge/base/envs/rstest2/lib/rstudio/RStudio.app/Contents/MacOS/resources/html/loading.html]; OCCURRED AT rstudio::core::Error rstudio::core::FilePath::openForRead(std::shared_ptr<std::istream> &) const src/cpp/shared_core/FilePath.cpp:1461; LOGGED FROM: void rstudio::desktop::MainWindow::onLoadFinished(bool) src/cpp/desktop/DesktopMainWindow.cpp:714
```

![Screen Shot 2022-03-25 at 8 11 18 PM](https://user-images.githubusercontent.com/1608317/160216602-117df9df-1c22-4543-8f91-b7b05c8980d2.png)

izahn commented:
I think this is ready for review. We have working Linux and OSX builds, and I think that is worth merging. 

There are two potential issues:

1. There is currently no Windows build. Adding a windows version may be more difficult (at least it is for me, I have no windows machine) but can be added later.
2. Despite working on it for a while in https://github.com/conda-forge/staged-recipes/pull/18401 I did not manage to install to standard directories as requested by @isuruf . I believe this does not affect the functionality, and I'm concerned that trying to "fix" it may break other things.

Despite these issues I think this is ready for review and merge. I'm sure I'm not alone in believing that having rstudio available via `conda` again will be a big improvement, even without Windows support at the moment.

isuruf commented:
I'm against merging this PR as is. The PR mentioned above is close to finishing. I looked at Gentoo and Gentoo has its own patch for fixing the layout. This may or may not be appropriate for us.
Also, I don't think your argument about breaking it in a different way is valid. If you look at the iterations we had, they were quite informative about how things broke. We can fix all those issues and possibly send the patch upstream for inclusion.

izahn commented:
Thanks @isuruf I appreciate your feedback.

> I looked at Gentoo and Gentoo has its own patch for fixing the layout.

That sounds potentially very promising, do you have a link handy?

> Also, I don't think your argument about breaking it in a different way is valid. If you look at the iterations we had, they were quite informative about how things broke. We can fix all those issues and possibly send the patch upstream for inclusion.

Maybe, but I'm unconvinced. My position here is pretty standard: this is not broken so we should not try to fix it. Attempting to do so does not improve the utility of the software and has the potential to break things.

izahn commented:
> The PR mentioned above is close to finishing. <SNIP> We can fix all those issues and possibly send the patch upstream for inclusion.

I would be much more comfortable doing this the other way around, i.e., submitting patches upstream where they can be reviewed and tested, and then coming back here to package it. But I don't have time for this and I don't consider the current version here broken, so this is another thing that I would be happy to see someone else do, but won't be doing myself.

izahn commented:
Sorry for the flurry of replies @isuruf , just one more thing:

> I'm against merging this PR as is. The PR mentioned above is close to finishing. 

It would be helpful to me if you would articulate why this matters to you and how you think it improves things. From my perspective this is entirely an aesthetics / style issue, it has zero bearing on how the software functions or how the software integrates into the larger conda-forge system. If I'm wrong about that it would help me if you would take a moment to explain why.

izahn commented:
Anything further here @isuruf ? 

I renew my request for review and merge @conda-forge/staged-recipes

isuruf commented:
> It would be helpful to me if you would articulate why this matters to you and how you think it improves things. From my perspective this is entirely an aesthetics / style issue, it has zero bearing on how the software functions or how the software integrates into the larger conda-forge system. If I'm wrong about that it would help me if you would take a moment to explain why.

No, it has consequences to how the software integrates to the ecosystem. When we go against the standard layout in unix, we run into lots of issues and accrue technical debt. For eg: `r-base` being installed into `$PREFIX/lib/R` instead of `$PREFIX/` has led us to use hacks like https://github.com/conda-forge/r-gmp-feedstock/blob/main/recipe/meta.yaml#L18-L20 all over the codebase. On windows, we use `$PREFIX/Library` instead of just `$PREFIX` which makes it so much harder to be operating system neutral.

> That sounds potentially very promising, do you have a link handy?

See https://gitweb.gentoo.org/repo/gentoo.git/tree/sci-mathematics/rstudio/files/rstudio-1.3.959-paths.patch

izahn commented:
> No, it has consequences to how the software integrates to the ecosystem. When we go against the standard layout in unix, we run into lots of issues and accrue technical debt. For eg: `r-base` being installed into `$PREFIX/lib/R` instead of `$PREFIX/` has led us to use hacks like https://github.com/conda-forge/r-gmp-feedstock/blob/main/recipe/meta.yaml#L18-L20 all over the codebase.

I see the issue with `r-base`, but that doesn't apply here at all since no shared libraries are included in the `rstudio-desktop` package at all. Are there other problems that do pertain to this package? If not then this really does come down to a style issue and I'm still not willing to go hacking around the code base just for aesthetics or style.

izahn commented:
@isuruf is there anything further here? Will you please withdraw your objection to the PR so we can move forward? 

I've made a substantial and good-faith effort to accommodate your request, but at the end of the day I still don't think it's a good idea. As the one volunteering to maintain this package I'm not willing to accept the added maintenance burden and risk. I don't think this package should be held up any longer because of this.

h-vetinari commented:
> I see the issue with `r-base`, but that doesn't apply here at all since no shared libraries are included in the `rstudio-desktop` package at all.

I think the larger point was about inconsistency creating subtle divergences, bugs, and ultimately lots of cruft & technical debt. I don't think anyone can fully predict all the ways we might be hit by this (both in how bad or irrelevant it will be), but the concern looks very valid to me.

> Will you please withdraw your objection to the PR so we can move forward?
> 
> I've made a substantial and good-faith effort to accommodate your request, but at the end of the day I still don't think it's a good idea.

I get how frustrating this is (I started https://github.com/conda-forge/staged-recipes/pull/13760 over a year ago), but that doesn't make this a reasonable demand either...

> As the one volunteering to maintain this package I'm not willing to accept the added maintenance burden and risk.

... because the point is exactly that the putative costs of this lack of consistency will be borne not by the package itself but by the ecosystem.

izahn commented:
> > I see the issue with `r-base`, but that doesn't apply here at all since no shared libraries are included in the `rstudio-desktop` package at all.
> 
> I think the larger point was about inconsistency creating subtle divergences, bugs, and ultimately lots of cruft & technical debt. I don't think anyone can fully predict all the ways we might be hit by this (in both directions, i.e. how bad or irrelevant it will be), but the concern looks very valid to me.

Maybe, but this is all very speculative and fuzzy. Maybe and some point in the future this will cause some unspecified problem. In contrast the requested change will for sure impose an additional maintenance cost right now. 

> 
> > Will you please withdraw your objection to the PR so we can move forward?
> > I've made a substantial and good-faith effort to accommodate your request, but at the end of the day I still don't think it's a good idea.
> 
> I get how frustrating this is (I started #13760 over a year ago), but that doesn't make this a reasonable demand either...

This is not a demand but a politely worded request :-)

> 
> > As the one volunteering to maintain this package I'm not willing to accept the added maintenance burden and risk.
> 
> ... because the point is exactly that the putative costs of this lack of consistency will be borne not by the package itself but by the ecosystem.

Maybe, but only in some vague, speculative, and specified way.  That's just not good enough to justify the requested change.

h-vetinari commented:
> > ... because the point is exactly that the putative costs of this lack of consistency will be borne not by the package itself but by the ecosystem.
> 
> Maybe, but only in some vague, speculative, and specified way. That's just not good enough to justify the requested change.

I think that's the crux here - you think the expected cost (probability * impact; over time) is low enough to justify moving forward, Isuru thinks it's high enough not to move forward as-is.

As everyone, you both have a different set of backgrounds and experiences that leads you to your respective conclusions; acknowledging that suggests being careful with statements like "just not good enough", That's already a good approach in general, but in particular when the other person is very knowledgeable about the subject matter (with several year's worth of work in every corner of conda-forge - consistently including contributions for the hardest problems at the time - Isuru's estimation is not exactly without foundation).

izahn commented:
Anything further here @isuruf ? Will you please drop your objection to this PR in light of the fact that it does not ship any shared libraries?

izahn commented:
@conda-forge-admin please restart ci

izahn commented:
@isuruf checking in to see what your current position is on this.

izahn commented:
Anybody else in @conda-forge/staged-recipes want to weigh in? I'd sure like to get this out there for the community.

isuruf commented:
My view is that this should not be merged as is, but that's just me. I'll defer to the others on what to do.

BastianZim commented:
Two cents from a reviewer: As conda-forge and the overall conda ecosystem try to be as "plug-and-play" and easy to use as possible, I would not vote in favour of merging this due to the concerns outlined by Isuru above. I would rather stay with the line of keeping everything as compatible and safe to use as possible since conda is sometimes used by an audience that will not be familiar with the potential difficulties imposed by some of the outlines mentioned above.
I fully understand, though that this is a frustrating experience and thanks, Ista, for coming this far! 
 
As a suggestion: What might be possible though, is a vote by core as these questions are generally deferred there. Alternatively, due to the overlap between the R and the conda-forge ecosystem: Do we know if some of the problems could be solved upstream, as suggested in https://github.com/conda-forge/staged-recipes/pull/18320#issuecomment-1090822565 ? They might be able to help solve some of the issues mentioned here.

BastianZim commented:
> Which potential difficulties specifically?

Those mentioned by Isuru above. Currently, they do only affect maintainers but since rstudio is such a cornerstone of the ecosystem and used widely, any changes later will need to be adopted by a wide userbase which is why I would prefer getting this really right before "unleashing" it, compared to a package that is only used by a niche audience.
It might probably be overkill to be this cautious though which is why I said only my two cents and would defer this to core. 😄

> Yes, working with rstudio upstream is something we should do. 

Would it be possible to make PRs for that and then add them as patches here until they are upstreamed? You mentioned your time constraint but maybe at least filing issues would help since it seems like other package managers have similar issues so maybe someone else in the community has an approach.

BastianZim commented on recipes/rstudio/meta.yaml:
```diff
@@ -0,0 +1,177 @@
+        - sh -c "xvfb-run -s '-screen 0 1200x800x24' ${PREFIX}/bin/rstudio --no-sandbox --run-diagnostics; cat ~/rstudio-diagnostics/diagnostics-report.txt"  # [linux]
+        - test -f ${PREFIX}/lib/rstudio/RStudio.app/Contents/MacOS/RStudio  # [osx]
+
+about:
+  home: www.rstudio.com/products/rstudio/
+  license: GPL-3.0-or-later
+  license_file: COPYING
+  doc_url: https://github.com/rstudio/rstudio
+  summary: RStudio is an integrated development environment (IDE) for R
```
You have two abouts. Maybe leftovers from the multi-output version?

h-vetinari replied:
I vote to keep the multi-output structure. Once we figure out what to do with pam we could/should also package rstudio-server in the same recipe, and keeping the multi-output from the get-go would avoid the indentation-churn.

BastianZim commented on recipes/rstudio/meta.yaml:
```diff
@@ -0,0 +1,177 @@
+  dev_url: https://github.com/rstudio/rstudio
+
+  
+extra:
```
I would add `feedstock-name` here as well, in case you keep this multi outpu.

izahn replied:
Done in https://github.com/izahn/staged-recipes/commit/380b10ffcb98a5e85dbf3016b85638878c079d0d , thank you!

izahn commented:
> > Which potential difficulties specifically?
> 
> Those mentioned by Isuru above. Currently, they do only affect maintainers but since rstudio is such a cornerstone of the ecosystem and used widely, any changes later will need to be adopted by a wide userbase which is why I would prefer getting this really right before "unleashing" it, compared to a package that is only used by a niche audience. It might probably be overkill to be this cautious though which is why I said only my two cents and would defer this to core.

Rstudio is an important package from the user perspective, but not really from the maintainer perspective. It is way out at the tip of the dependency tree; for example, probably nothing will ever depend on it.

There is a lack of agreement about what "right" is. I think it is right to use the build process tested by the upstream devs, patching it only when required.

> 
> > Yes, working with rstudio upstream is something we should do.
> 
> Would it be possible to make PRs for that and then add them as patches here until they are upstreamed? You mentioned your time constraint but maybe at least filing issues would help since it seems like other package managers have similar issues so maybe someone else in the community has an approach.

Yes it is possible, but my motivation is very low, honestly I probably will not find time to do it. Happy if someone else wants to do it though.

ocefpaf commented:
@izahn we, core members, just had a quick chat about this PR in out meeting and here is the gist of it:

- @isuruf's concern are important and we should work to address them, either with more patching (yey more technical debit), or working with upstream;
- AnacondaInc will probably pull their Rstudio package from defaults b/c the problems with this approach makes it unusable there; 
- With all that said, we can merge this, skip Windows (as you are already doing), and keep track of the problems so we can address them in the future.


What is the status on macOS? Do you want to skip that one or try to fix the CI?

BTW, big thanks to you, all that worked in the PR and review it. This is an important package for the R community, very difficult to build, and hopefully we will make some users happy by putting something that works out there.

izahn commented:
Awesome, thanks @ocefpaf and everyone who worked on this!

The OSX package was working, I'll get it building again later today.

ocefpaf commented:
> The OSX package was working, I'll get it building again later today.

Please ping me when you are done.

ocefpaf commented:
> Should I add @conda-forge/r to the maintainer list?

Let's "cash in" this into a feedstock and they can add the team there if they want to maintain this too.

gdevenyi commented on recipes/rstudio/meta.yaml:
```diff
@@ -0,0 +1,173 @@
+  version: {{ version }}
+
+source:
+  url: "https://github.com/rstudio/rstudio/archive/refs/tags/v{{ version }}+485.tar.gz"
```
I think the +485 needs to go into the version string, as it will change (its the build count on their dailies before they get a release).

gdevenyi replied:
https://github.com/conda-forge/rstudio-feedstock/pull/1

izahn replied:
Thank you!
